packages/styled-base/types/index.d.ts

```diff
@@ -24,127 +23,88 @@ export {
   ObjectInterpolation
 } from '@emotion/serialize'
 
-export { ComponentSelector, Interpolation }
-
-type JSXInEl = JSX.IntrinsicElements
-
-export type WithTheme<P, T> = P extends { theme: infer Theme }
-  ? P & { theme: Exclude<Theme, undefined> }
-  : P & { theme: T }
+export { ComponentSelector, Interpolation, PropsOf, Omit }
 
 export interface StyledOptions {
   label?: string
   shouldForwardProp?(propName: string): boolean
   target?: string
 }
 
-export interface StyledComponent<InnerProps, StyleProps, Theme extends object>
-  extends React.SFC<InnerProps & StyleProps & { theme?: Theme }>,
+/**
+ * @typeparam ComponentProps  Props which will be included when withComponent is called
+ * @typeparam SpecificComponentProps  Props which will *not* be included when withComponent is called
+ */
+export interface StyledComponent<ComponentProps, SpecificComponentProps = {}>
+  extends React.FC<ComponentProps & SpecificComponentProps>,
     ComponentSelector {
-  /**
-   * @desc this method is type-unsafe
-   */
-  withComponent<NewTag extends keyof JSXInEl>(
-    tag: NewTag
-  ): StyledComponent<JSXInEl[NewTag], StyleProps, Theme>
-  withComponent<Tag extends React.ComponentType<any>>(
+  withComponent<C extends React.ComponentType<React.ComponentProps<C>>>(
+    component: C
+  ): StyledComponent<ComponentProps & PropsOf<C>>
+  withComponent<Tag extends keyof JSX.IntrinsicElements>(
     tag: Tag
-  ): StyledComponent<PropsOf<Tag>, StyleProps, Theme>
-}
-
-type ReactClassPropKeys = keyof React.ClassAttributes<any>
-
-interface CreateStyledComponentBaseThemeless<InnerProps, ExtraProps> {
-  <
-    StyleProps extends Omit<
-      Overwrapped<InnerProps, StyleProps>,
-      ReactClassPropKeys
-    > = Omit<InnerProps & ExtraProps, ReactClassPropKeys>,
-    Theme extends object = object
-  >(
-    ...styles: Array<Interpolation<WithTheme<StyleProps, Theme>>>
-  ): StyledComponent<InnerProps, StyleProps, Theme>
-  <
-    StyleProps extends Omit<
-      Overwrapped<InnerProps, StyleProps>,
-      ReactClassPropKeys
-    > = Omit<InnerProps & ExtraProps, ReactClassPropKeys>,
-    Theme extends object = object
-  >(
-    template: TemplateStringsArray,
-    ...styles: Array<Interpolation<WithTheme<StyleProps, Theme>>>
-  ): StyledComponent<InnerProps, StyleProps, Theme>
+  ): StyledComponent<ComponentProps, JSX.IntrinsicElements[Tag]>
 }
 
-interface CreateStyledComponentBaseThemed<
-  InnerProps,
-  ExtraProps,
-  StyledInstanceTheme extends object
+/**
+ * @typeparam ComponentProps  Props which will be included when withComponent is called
+ * @typeparam SpecificComponentProps  Props which will *not* be included when withComponent is called
+ * @typeparam StyleProps  Params passed to styles but not exposed as React props. These are normally library provided props
```

Comment:
would it even be possible to build a wrapper over `styled` API or smth that would actually use this? 🤔

Reply:
I could export a different type from the `styled` package which is compatible with the old syntax?

Reply:
Hm, not sure if we are on the same page - I'm wondering purely about how useful that third param is. Like I don't see immediately how one could use it. Is there any test covering this? Could you point me to it?

Reply:
Here is where it's used

<img width="377" alt="image" src="https://user-images.githubusercontent.com/453152/67680632-97f88c00-f9c6-11e9-86f7-934d0e2ec70a.png">

From the outside `{ theme?: Theme }` will be included with withComponent, but the intrinsic props won't, then `{ theme: Theme }` is not optional when consuming props inside the interpolations.